// Use make all to build python classes.

package nn;

message LearningParams {
  required float l1_wt = 1 [default = 0.0001] ; 
  required float l2_wt = 2 [default = 0.0001] ; 
  required float epsilon = 3 [default = 0.001] ; 
  required float epsilon_anneal_rate = 4 [default = 0.99] ; 
  required float momentum  = 5 [default = 0.9] ; 
  required int32 start_epoch = 6 [default = 0] ;
  required int32 end_epoch = 7 [default = 100000] ;
  required float dropout_rate = 8 [default = 0] ;
  required float wt_norm_constraint = 9 [default = 0] ;
  required float ada_lambda = 10 [default = 0.95] ;
}


message InitParams {
  required float wt_sigma = 1 [default = 3.0] ;
  required float biases_min = 2 [default = -5.0] ;
  required float biases_max = 3 [default = -3.0] ;
}

message Layer {
  required string name = 1;
  // ignored in mixture of softmax
  required int32 input_dim = 2;
  required int32 num_units = 3;
  enum LayerType {
    LINEAR = 0;
    SIGMOID = 1;
    RELU= 2;
    SOFTMAX = 3;
    MULTI_SOFTMAX = 4;
  }
  required LayerType type = 4;
  required InitParams init_params = 5;
  repeated LearningParams learning_schedule = 6 ;
  optional int32 num_softmaxes = 7;
}

message nn {
  repeated Layer layers = 1;
}
